torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py

```diff
@@ -1398,11 +1394,12 @@ def _free_full_params(self,
     for p in full_params:
       if p._has_full_param:
         # free the original full parameter
-        p.data = self._dummy_data_placeholder
+        with torch.autograd._unsafe_preserve_version_counter(p):
+          p.set_(self._dummy_data_placeholder)
         p._has_full_param = False
 
     if apply_opt_barrier:
-      self._apply_opt_barrier_to_params_and_tensors(full_params, sharded_params,
+      self._apply_opt_barrier_to_params_and_tensors([p for p in full_params if p._has_full_param], sharded_params,
```

Comment:
@ronghanghu Wonder why we want to `_apply_opt_barrier_to_params_and_tensors`  to p even if it doesn't hold the full param?

Reply:
@alanwaketan Yeah, you're right! It's probably not needed here on those `p` without full param. (I probably introduced several unnecessary optimization barriers in the implementation earlier.)